File: ts/src/adapters/omp.ts
import { register } from '../registry.js'
import type { Adapter, BuildCommand, ParsedOutput, RunSpec, SubprocOutcome } from '../base.js'
import { finalizeCommand, validateRunSpec } from '../base.js'

// Oh My Pi (omp) `--print --mode json` writes one JSON object per stdout line.
// An assistant message carries usage { input, output, cacheRead, cacheWrite,
// totalTokens, cost: { total, ... } }. Qualified with 18.1.10, a cycle looks like:
//
//   message_end (assistant, final usage) → turn_end (same usage again)
//   → agent_end { messages: [...] }
//
// `agent_end.messages` is authoritative for its completed cycle. A stream cut
// before it leaves `message_end` records; `turn_end` is only a fallback for a
// turn that never produced an assistant `message_end`. `message_update`
// partials never carry usage worth counting.
//
// Exit status is the real process status: a provider error surfaces in the
// records (`error` / `stopReason`), and omp may still exit 0.

type JsonObject = Record<string, unknown>

interface Usage {
  tokensIn: number | null
  tokensOut: number | null
  costUsd: number | null
}

function isJsonObject(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function add(total: number | null, value: number | null): number | null {
  return value === null ? total : (total ?? 0) + value
}

function merge(into: Usage, from: Usage): void {
  into.tokensIn = add(into.tokensIn, from.tokensIn)
  into.tokensOut = add(into.tokensOut, from.tokensOut)
  into.costUsd = add(into.costUsd, from.costUsd)
}

/**
 * Adds one assistant message's usage. Each metric counts independently and
 * only when well-typed: token counts as nonnegative integers, cost.total as a
 * finite nonnegative number; anything else leaves that metric untouched.
 * Returns whether `message` was an assistant message at all.
 */
function addAssistantUsage(into: Usage, message: unknown): boolean {
  if (!isJsonObject(message) || message['role'] !== 'assistant') return false
  const usage = message['usage']
  if (!isJsonObject(usage)) return true
  const input = usage['input']
  const output = usage['output']
  const cost = usage['cost']
  const total = isJsonObject(cost) ? cost['total'] : undefined
  if (typeof input === 'number' && Number.isInteger(input) && input >= 0) {
    into.tokensIn = add(into.tokensIn, input)
  }
  if (typeof output === 'number' && Number.isInteger(output) && output >= 0) {
    into.tokensOut = add(into.tokensOut, output)
  }
  if (typeof total === 'number' && Number.isFinite(total) && total >= 0) {
    into.costUsd = add(into.costUsd, total)
  }
  return true
}

function parseOmpEvents(stdout: string): ParsedOutput {
  const records: JsonObject[] = []
  for (const line of stdout.split('\n')) {
    const s = line.trim()
    if (!s) continue
    let parsed: unknown
    try {
      parsed = JSON.parse(s)
    } catch {
      continue
    }
    if (isJsonObject(parsed)) records.push(parsed)
  }
  if (records.length === 0) return { costUsd: null, tokensIn: null, tokensOut: null, raw: null }

  const totals: Usage = { tokensIn: null, tokensOut: null, costUsd: null }
  let pending: Usage = { tokensIn: null, tokensOut: null, costUsd: null }
  // The current turn already contributed an assistant `message_end`; its
  // `turn_end` repeats that usage and must not be counted again.
  let turnCounted = false

  for (const record of records) {
    switch (record['type']) {
      case 'message_end':
        if (addAssistantUsage(pending, record['message'])) turnCounted = true
        break
      case 'turn_end':
        if (!turnCounted) addAssistantUsage(pending, record['message'])
        turnCounted = false
        break
      case 'agent_end': {
        const messages = record['messages']
        if (Array.isArray(messages)) {
          for (const message of messages) addAssistantUsage(totals, message)
        } else {
          merge(totals, pending)
        }
        pending = { tokensIn: null, tokensOut: null, costUsd: null }
        turnCounted = false
        break
      }
    }
  }
  // Incomplete final cycle: keep what its message_end / turn_end records reported.
  merge(totals, pending)

  return { costUsd: totals.costUsd, tokensIn: totals.tokensIn, tokensOut: totals.tokensOut, raw: records }
}

const ompAdapter: Adapter = {
  name: 'omp',
  instructionsFilename: 'AGENTS.md',
  defaultModel: 'sonnet',
  permissionBypassArgs: ['--auto-approve'],
  configHomeEnv: 'PI_CODING_AGENT_DIR',
  configFileFlag: '--config',

  buildCommand(spec: RunSpec): BuildCommand {
    const validated = validateRunSpec(this, spec)
    const { model, permissionArgs, configHome, configArgs } = validated
    const args = ['--print', '--mode', 'json', '--no-session', '--model', model]
    // An explicit agent dir must win over an inherited OMP_PROFILE, which would
    // otherwise ignore that dir and select state under the config root.
    if (configHome !== null) args.push('--profile', 'default')
    args.push(...permissionArgs, ...configArgs, '--', spec.prompt)
    return finalizeCommand(this, spec, validated, { cmd: 'omp', args })
  },

  parseOutput(_spec: RunSpec, outcome: SubprocOutcome): ParsedOutput {
    return parseOmpEvents(outcome.stdout)
  },
}

ompAdapter.installMeta = {
  packageManager: 'brew',
  installCommand: ['brew', 'install', 'can1357/tap/omp'],
  updateCommand: ['brew', 'upgrade', 'omp'],
  versionCommand: ['omp', '--version'],
}

register('omp', ompAdapter)

File: src/harness/adapters/omp.py
"""omp adapter — invokes the Oh My Pi `omp` CLI in headless print mode.

`omp --print --mode json` emits one JSON object per stdout line. Qualified
with OMP 18.1.10, the stream carries, per agent cycle:
  - `message_end`   — a completed message (assistant messages carry `usage`)
  - `turn_end`      — the same assistant message again (may be duplicated)
  - `agent_end`     — `messages`: the full transcript of the cycle
`message_update` records are streaming partials and carry no usable totals.

Each AssistantMessage has:
    usage: { input, output, cacheRead, cacheWrite, totalTokens,
             cost: { input, output, cacheRead, cacheWrite, total } }

Every `agent_end` with a `messages` array is authoritative for its cycle; cycles
are aggregated. For a cycle that never reached `agent_end` (timeout, kill) the
`message_end` assistant records are summed, falling back to `turn_end` only for
turns that produced no `message_end`.

Provider errors are reported inside the JSON stream (`error` / `stopReason`
fields on the message) and omp may still exit 0; `RunResult.exit_code` is the
process status, and the native records are preserved verbatim in `raw`.
"""
from __future__ import annotations

import json
import math

from harness._subproc import SubprocOutcome
from harness.base import Adapter, BuildCommand, InstallMeta, ParsedOutput, RunSpec

_Totals = tuple[int | None, int | None, float | None]
_NO_TOTALS: _Totals = (None, None, None)


class OmpAdapter(Adapter):
    name = "omp"
    instructions_filename = "AGENTS.md"
    permission_bypass_args = ("--auto-approve",)
    config_home_env = "PI_CODING_AGENT_DIR"
    config_file_flag = "--config"
    install_meta = InstallMeta(
        package_manager="brew",
        install_command=("brew", "install", "can1357/tap/omp"),
        update_command=("brew", "upgrade", "omp"),
        version_command=("omp", "--version"),
    )

    DEFAULT_MODEL = "sonnet"

    def build_command(self, spec: RunSpec) -> BuildCommand:
        resolved = self.resolve_run_spec(spec)
        args = ["--print", "--mode", "json", "--no-session", "--model", resolved.model]
        if spec.config_home is not None:
            # An explicit agent dir must win over a named profile inherited
            # from the caller's environment.
            args += ["--profile", "default"]
        args += [*resolved.permission_args, *resolved.native_args, *resolved.config_args, "--", spec.prompt]
        return self.finalize_command(spec, cmd="omp", args=args)

    def parse_output(self, spec: RunSpec, outcome: SubprocOutcome) -> ParsedOutput:
        tokens_in, tokens_out, cost, raw = _parse_omp_records(outcome.stdout)
        return {"cost_usd": cost, "tokens_in": tokens_in, "tokens_out": tokens_out, "raw": raw}


def _parse_omp_records(stdout: str) -> tuple[int | None, int | None, float | None, list | None]:
    records: list[dict] = []
    for line in stdout.split("\n"):
        line = line.strip()
        if not line:
            continue
        try:
            record = json.loads(line)
        except json.JSONDecodeError:
            continue
        if isinstance(record, dict):
            records.append(record)
    if not records:
        return None, None, None, None

    totals = _NO_TOTALS
    pending = _NO_TOTALS
    turn_seen = False  # a message_end assistant record was seen for the current turn
    for record in records:
        kind = record.get("type")
        if kind == "message_end":
            usage = _assistant_usage(record.get("message"))
            if usage is not None:
                pending = _add(pending, usage)
                turn_seen = True
        elif kind == "turn_end":
            if not turn_seen:
                usage = _assistant_usage(record.get("message"))
                if usage is not None:
                    pending = _add(pending, usage)
            turn_seen = False
        elif kind == "agent_end":
            messages = record.get("messages")
            if isinstance(messages, list):
                for message in messages:
                    usage = _assistant_usage(message)
                    if usage is not None:
                        totals = _add(totals, usage)
            else:
                totals = _add(totals, pending)
            pending = _NO_TOTALS
            turn_seen = False
    totals = _add(totals, pending)
    return totals[0], totals[1], totals[2], records


def _assistant_usage(message: object) -> _Totals | None:
    """Validated `(input, output, cost.total)` of an assistant message; None if not one."""
    if not isinstance(message, dict) or message.get("role") != "assistant":
        return None
    usage = message.get("usage")
    if not isinstance(usage, dict):
        return _NO_TOTALS
    cost = usage.get("cost")
    return (
        _token_count(usage.get("input")),
        _token_count(usage.get("output")),
        _cost(cost.get("total")) if isinstance(cost, dict) else None,
    )


def _token_count(value: object) -> int | None:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    if value < 0 or (isinstance(value, float) and not value.is_integer()):
        return None
    return int(value)


def _cost(value: object) -> float | None:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    if value < 0 or not math.isfinite(value):
        return None
    return float(value)


def _add(a: _Totals, b: _Totals) -> _Totals:
    """Per-metric sum; a metric stays None until some record supplies a valid value."""
    return (
        a[0] if b[0] is None else (a[0] or 0) + b[0],
        a[1] if b[1] is None else (a[1] or 0) + b[1],
        a[2] if b[2] is None else (a[2] or 0) + b[2],
    )
